rollup/circuitcapacitychecker/libzkp/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+#![feature(once_cell)]
+
+pub mod checker {
+    use crate::utils::{c_char_to_vec, vec_to_c_char};
+    use libc::c_char;
+    use prover::zkevm::{CircuitCapacityChecker, RowUsage};
+    use serde_derive::{Deserialize, Serialize};
+    use std::cell::OnceCell;
+    use std::collections::HashMap;
+    use std::panic;
+    use std::ptr::null;
+    use types::eth::BlockTrace;
+
+    #[derive(Debug, Clone, Deserialize, Serialize)]
+    pub struct RowUsageResult {
+        pub acc_row_usage: Option<RowUsage>,
+        pub tx_row_usage: Option<RowUsage>,
+        pub error: Option<String>,
+    }
+
+    static mut CHECKERS: OnceCell<HashMap<u64, CircuitCapacityChecker>> = OnceCell::new();
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn init() {
+        env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("debug"))
+            .format_timestamp_millis()
+            .init();
+        let checkers = HashMap::new();
+        CHECKERS.set(checkers).unwrap();
+    }
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn new_circuit_capacity_checker() -> u64 {
+        let id = CHECKERS.get_mut().unwrap().len() as u64;
+        let checker = CircuitCapacityChecker::new();
+        CHECKERS.get_mut().unwrap().insert(id, checker);
+        id
+    }
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn reset_circuit_capacity_checker(id: u64) {
+        CHECKERS.get_mut().unwrap().get_mut(&id).unwrap().reset()
+    }
+
+    /// # Safety
+    #[no_mangle]
+    pub unsafe extern "C" fn apply_tx(id: u64, tx_traces: *const c_char) -> *const c_char {
+        let tx_traces_vec = c_char_to_vec(tx_traces);
+        let traces = serde_json::from_slice::<BlockTrace>(&tx_traces_vec).unwrap();
+        let result = panic::catch_unwind(|| {
+            CHECKERS
+                .get_mut()
+                .unwrap()
+                .get_mut(&id)
+                .unwrap()
+                .estimate_circuit_capacity(&[traces])
+                .unwrap()
+        });
+        let r = match result {
+            Ok((acc_row_usage, tx_row_usage)) => {
+                log::debug!(
+                    "acc_row_usage: {:?}, tx_row_usage: {:?}",
```

Comment:
I suggest including the `id` here, otherwise these logs might not be very useful

Reply:
fixed in https://github.com/scroll-tech/go-ethereum/pull/321/commits/714afd721af62aa19b2725994f8154509950f629